django/test/testcases.py

```diff
@@ -150,6 +150,48 @@ def __call__(self):
         )
 
 
+class _AssertSignalSentContext(object):
+    def __init__(self, test_case, signal, required_kwargs=None):
+        self.test_case = test_case
+        self.signal = signal
+        self.required_kwargs = required_kwargs
+
+    def _listener(self, sender, **kwargs):
+        self.signal_sent = True
+        self.received_kwargs = kwargs
+        self.sender = sender
+
+    def __enter__(self):
+        self.signal_sent = False
+        self.received_kwargs = {}
+        self.sender = None
+        self.signal.connect(self._listener)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.signal.disconnect(self._listener)
+        if not self.signal_sent:
+            self.test_case.fail('Signal was not sent.')
```

Comment:
It could be helpful to add the expected signal's name in these messages as the error message doesn't look particularly friendly to debug, particularly if multiple context managers are used in a single `with`.
```
======================================================================
FAIL: test_login (auth_tests.test_signals.SignalTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/auth_tests/test_signals.py", line 49, in test_login
    self.assertIn('request', self.login_failed[0])
  File "/opt/python3.5.2/lib/python3.5/contextlib.py", line 66, in __exit__
    next(self.gen)
  File "/home/tim/code/django/django/test/testcases.py", line 699, in assertSignalSent
    yield cm
  File "/home/tim/code/django/django/test/testcases.py", line 174, in __exit__
    self.test_case.fail('Signal was not sent.')
AssertionError: Signal was not sent.
```